[WIP] Add aot_inductor as a test backend for benchmarking

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104977

Summary: Add an aot_inductor backend (Export+AOTInductor) in the benchmarking harness. Note it is not a dynamo backend.
* Moved files from torch/_inductor/aot_inductor_include to torch/csrc/inductor as a more standard way for exposing headers
* Created a caching function in benchmarks/dynamo/common.py for compiling, loading and caching the .so file

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @anijain2305

Differential Revision: [D47410386](https://our.internmc.facebook.com/intern/diff/D47410386)

desertfire committed:
[WIP] Add aot_inductor as a test backend for benchmarking

Summary: Call Export+AOTInductor to generate a .so file, and load that .so back into python for benchmarking.

[ghstack-poisoned]

desertfire committed:
Update on "[WIP] Add aot_inductor as a test backend for benchmarking"


Summary: Call Export+AOTInductor to generate a .so file, and load that .so back into python for benchmarking.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 anijain2305

[ghstack-poisoned]

chenyang78 commented on torch/csrc/inductor/aot_inductor_interface.h:
```diff
@@ -1,8 +1,50 @@
-#include "aot_inductor_interface.h"
```
This probably wouldn't work for the internal uses, where the definitions of the interfaces are included in two places:

* the generated .so (via ```#include <aot_inductor_interface.h>```; and
* the runtime that includes <aot_inductor_interface.h>

I am thinking perhaps we could rename aot_inductor_interface.h to aot_inductor_interface_impl.h while keeping the original aot_inductor_interface.h, if we don't want to have any wrapper code with a .cpp suffix.

desertfire replied:
Good catch. I think in the generated .so, we only needs to include aot_inductor_model.h because that's that part we will generate code as its definition. aot_inductor_interface.h|cpp should be generated as a part of libtorch, and we will also need take care of the compatibility issues for them using the same mechanism we have discussed.

desertfire replied:
I probably should commit from fbcode to make sure things work there.

chenyang78 replied:
> Good catch. I think in the generated .so, we only needs to include aot_inductor_model.h because that's that part we will generate code as its definition. aot_inductor_interface.h|cpp should be generated as a part of libtorch, and we will also need take care of the compatibility issues for them using the same mechanism we have discussed.

Hmm, if the generated .so doesn't contain interface.h/cpp, how would the runtime dlopen this .so and call the C APIs?

chenyang78 replied:
> I probably should commit from fbcode to make sure things work there.

Yes, it's probably a good idea :)

ezyang commented:
I don't have any objection to having this be a benchmark configuration, though I thought we already had cpp wrapper backend. Are they redundant?

desertfire commented:
> I don't have any objection to having this be a benchmark configuration, though I thought we already had cpp wrapper backend. Are they redundant?

Comparing to cpp_wrapper, this PR can test the path from Export to AOTInductor, and then people can use it to improve the Export API/functionality/performance. [Here](https://hud.pytorch.org/benchmark/compilers?startTime=Wed%2C%2005%20Jul%202023%2012%3A58%3A38%20GMT&stopTime=Wed%2C%2012%20Jul%202023%2012%3A58%3A38%20GMT&granularity=hour&suite=torchbench&mode=inference&dtype=bfloat16&lBranch=gh/desertfire/191/head&lCommit=bf3c9f5231e330e588f9085b44660fa8bd4fe904&rBranch=main&rCommit=54e320d4d1654d45b65c3a5a1ba8fe35faf21460) is a dashboard run that I compare this new backend results with cpp-wrapper results. Once obvious issue there is we are still lagging in coverage there.

I will remove the change to `test.sh` when submitting this PR. Once we have a reasonable pass rate for the aot_inductor backend, we can replace cpp_wrapper with aot_inductor on the dashboard.

ezyang commented:
OK SGTM

desertfire committed:
Update on "[WIP] Add aot_inductor as a test backend for benchmarking"


Summary: Add an aot_inductor backend (Export+AOTInductor) in the benchmarking harness. Note it is not a dynamo backend.
* Moved files from torch/_inductor/aot_inductor_include to torch/csrc/inductor as a more standard way for exposing headers
* Created a caching function in benchmarks/dynamo/common.py for compiling, loading and caching the .so file

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 anijain2305

[ghstack-poisoned]

desertfire committed:
Update on "[WIP] Add aot_inductor as a test backend for benchmarking"


Summary: Add an aot_inductor backend (Export+AOTInductor) in the benchmarking harness. Note it is not a dynamo backend.
* Moved files from torch/_inductor/aot_inductor_include to torch/csrc/inductor as a more standard way for exposing headers
* Created a caching function in benchmarks/dynamo/common.py for compiling, loading and caching the .so file

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 anijain2305

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47410386).

desertfire committed:
Update on "[WIP] Add aot_inductor as a test backend for benchmarking"


Summary: Add an aot_inductor backend (Export+AOTInductor) in the benchmarking harness. Note it is not a dynamo backend.
* Moved files from torch/_inductor/aot_inductor_include to torch/csrc/inductor as a more standard way for exposing headers
* Created a caching function in benchmarks/dynamo/common.py for compiling, loading and caching the .so file

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 anijain2305

Differential Revision: [D47410386](https://our.internmc.facebook.com/intern/diff/D47410386)

[ghstack-poisoned]

desertfire committed:
Update on "[WIP] Add aot_inductor as a test backend for benchmarking"


Summary: Add an aot_inductor backend (Export+AOTInductor) in the benchmarking harness. Note it is not a dynamo backend.
* Moved files from torch/_inductor/aot_inductor_include to torch/csrc/inductor as a more standard way for exposing headers
* Created a caching function in benchmarks/dynamo/common.py for compiling, loading and caching the .so file

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 anijain2305

Differential Revision: [D47410386](https://our.internmc.facebook.com/intern/diff/D47410386)

[ghstack-poisoned]

desertfire commented:
close in favor of https://github.com/pytorch/pytorch/pull/105221
